pyflow/RCDT2/Nodes/core.py

```diff
@@ -0,0 +1,183 @@
+# SPDX-FileCopyrightText: Alliander N. V.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from PyFlow.Core import NodeBase, PinBase
+import rclpy
+from rclpy.node import Node
+from threading import Thread
+from time import sleep
+from logging import getLogger
+from inflection import underscore
+
+logger = getLogger(__name__)
+
+
+class PinManager:
+    def __init__(self):
+        self.input_pins: dict[str, PinBase] = {}
+        self.output_pins: dict[str, PinBase] = {}
+
+    def add_input_pin(self, pin_name: str, pin: PinBase) -> None:
+        self.input_pins[pin_name] = pin
+
+    def add_output_pin(self, pin_name: str, pin: PinBase) -> None:
+        self.output_pins[pin_name] = pin
+
+    def get_data(self, pin_name: str) -> object:
+        return self.input_pins[pin_name].getData()
+
+    def set_data(self, pin_name: str, data: object) -> object:
+        return self.output_pins[pin_name].setData(data)
+
+    def get_input_pin_names(self) -> list[str]:
+        return self.input_pins.keys()
+
+    def get_output_pin_names(self) -> list[str]:
+        return self.output_pins.keys()
+
+
+class PyflowRosBridge:
+    rclpy.init()
+    node = Node("pyflow")
+    Thread(target=rclpy.spin, args=[node], daemon=True).start()
+    node.get_logger().info("PyFlow ROS Node started!")
+
+
+class PyflowBase(NodeBase):
+    def __init__(self, name: str):
+        super().__init__(name)
+        self.pin_manager = PinManager()
+
+
+class PyflowComputer(PyflowBase):
+    def compute(self, *_args: any, **_kwargs: any) -> None:
+        raise NotImplementedError
+
+
+class PyflowExecutor(PyflowBase):
+    def execute(self, *_args: any, **_kwargs: any) -> None:
+        raise NotImplementedError
+
+
+class PyflowParallelExecutor(PyflowExecutor):
+    def __init__(self, name: str):
+        super().__init__(name)
+        self.parallel_executor_is_active = False
+
+    def execute_parallel(self, *_args: any, **_kwargs: any) -> None:
+        if self.parallel_executor_is_active:
+            logger.error("Executor is already working. Rejected.")
+            return
+        thread = Thread(target=self.execution_steps)
+        thread.start()
+
+    def execution_steps(self) -> None:
+        self.parallel_executor_is_active = True
+        self.execute()
+        self.parallel_executor_is_active = False
+
+
+class Service(PyflowParallelExecutor):
+    def __init__(self, name: str):
+        super().__init__(name)
+        self.service_type: type
+
+        self.createInputPin("In", "ExecPin", callback=self.execute_parallel)
+        self.exec_out: PinBase = self.createOutputPin("Out", "ExecPin")
+
+        for service_part in ["Request", "Response"]:
+            self.create_data_pins(service_part)
+
+        self.topic_name = underscore(self.service_type.__name__)
```

Comment:
topic_name klopt eigenlijk net niet, als ik de documentatie van ROS goed begrijp dan is dit de service_name

bron: https://docs.ros.org/en/foxy/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Services/Understanding-ROS2-Services.html#ros2-service-call